### ISSUE TITLE
Ensure it works with pgtk

### ISSUE THREAD

**Issue by meedstrom** [thblt/sway.el#3]
I'm using Sway with Xwayland disabled, because pgtk makes Emacs a native Wayland app.  I noticed that `sway-list-frames` relies on `sway-find-x-window-frame` which relies on X window IDs.  I wrote a workable alternative, though I still need to figure out how to uniquify frame names to be on the safe side.

You could check for `(eq window-system 'pgtk)` and conditionally use either this or the X function, because it's always better with truly unique ids.

By the way, the docstring for `sway-find-x-window-frame` is misleading.  It says the parameter WINDOW is an X window ID, which it is not, it is a hash table of a Sway window's properties.  This tripped me up while rewriting the function. ;-)

My Wayland version:

``` emacs-lisp
(defun sway-find-wayland-window-frame (window)
  "Return the Emacs frame corresponding to a Sway window WINDOW.
Does this by matching frame names, so you may want to ensure
they are always uniquified.

WINDOW is a hash table, typically one of the members of
`sway-list-windows'."
  (require 'map)
  (let ((name (map-elt window "name")))
    (--find (equal (frame-parameter it 'name) name)
            (frame-list))))
```

And the patched `sway-list-frames` for anyone reading (simply insert both snippets in your initfiles to cure the errors).

``` emacs-lisp
(require 'el-patch) ;; optional
(el-patch-defun sway-list-frames (&optional tree visible-only focused-only)
  "List all Emacs frames in TREE.

VISIBLE-ONLY and FOCUSED-ONLY select only frames that are,
respectively, visible and focused.

Return value is a list of (FRAME-OBJECT . SWAY-ID)"
  (unless tree (setq tree (sway-tree)))
  (let* ((wins (sway-list-windows tree visible-only focused-only)))
    (seq-filter (lambda (x) (car x))
                (-zip
                 (mapcar (if (eq window-system 'pgtk)
                             #'sway-find-wayland-window-frame
                           #'sway-find-x-window-frame)
                         wins)
                 (mapcar #'sway-get-id wins)))))
```

**Comment by thblt** [comment]
Thank you for the suggestion and the code samples, and apologies for the very late reply. 

I'll get to implement this at some point.

I'd like to find a better way to match Emacs frames with their corresponding sway windows than name matching.   Maybe there's some equivalent to `'window-id` I could use.

**Comment by nolanl** [comment]
Unfortunately 'window-id on pgtk is the virtual address of the GtkWindow in emacs' address space. I've got the same problem in a tool of my own, and I've not been able to come up with a better solution than name matching.

**Comment by thblt** [comment]
I've been making some progress on this issue, but it isn't working yet.  I implemented your function along with a dispatcher to pick the x or wayland function.

It isn't done yet, because I then tried to improve frame matching by giving them temporary unique names, and hit [a possible Emacs/pgtk bug](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=58945).  I'm waiting for feedback on that report to find a way to move on.  Depending on the resolution, we may either stay with your basic approach, or use a more aggressive way to uniquify frame names, like hooking into `window-state-change-hook` to dynamically keep frame names unique (for example, if two frames display the same buffer on their active window, we could just rename them "buffer-name (1)" and "buffer-name (2)").

btw, the Sway tree gives us the pid of the process owning each window, so there's no need to be concerned with collision with other programs or Emacs instances.  I've slightly modified `sway-list-windows` to add pid filtering.
